symbolic-unwind/src/base.rs

```diff
@@ -26,67 +98,104 @@ pub trait Address:
     + Sub<Output = Self>
     + Rem<Output = Self>
     + Copy
-    + std::fmt::Debug
+    + Sized
+    + Debug
 {
+    /// The number of bytes that need to be read to produce one value of this type.
+    const WIDTH: usize;
+    /// Attempt to read a value of this type from a slice of bytes.
+    ///
+    /// May fail if an invalid byte is encountered or there are not enough bytes in the slice.
+    fn read_bytes<E: Endianness>(bytes: &[u8], endian: E) -> Option<Self>;
 }
 
-impl ReadBytes for u8 {
+impl RegisterValue for u8 {
     const WIDTH: usize = 1;
-    fn read_bytes<B: ByteOrder>(bytes: &[u8]) -> Option<Self> {
+    fn read_bytes<E: Endianness>(bytes: &[u8], _endian: E) -> Option<Self> {
         bytes.first().copied()
     }
 }
 
-impl ReadBytes for u16 {
+impl RegisterValue for u16 {
     const WIDTH: usize = 2;
-    fn read_bytes<B: ByteOrder>(bytes: &[u8]) -> Option<Self> {
-        (bytes.len() >= Self::WIDTH).then(|| B::read_u16(bytes))
+    fn read_bytes<E: Endianness>(bytes: &[u8], endian: E) -> Option<Self> {
+        let bytes: &[u8; Self::WIDTH] = bytes[..Self::WIDTH].try_into().ok()?;
+        if endian.is_big_endian() {
+            Some(Self::from_be_bytes(*bytes))
+        } else {
+            Some(Self::from_le_bytes(*bytes))
+        }
     }
 }
 
-impl ReadBytes for u32 {
+impl RegisterValue for u32 {
     const WIDTH: usize = 4;
-    fn read_bytes<B: ByteOrder>(bytes: &[u8]) -> Option<Self> {
-        (bytes.len() >= Self::WIDTH).then(|| B::read_u32(bytes))
+    fn read_bytes<E: Endianness>(bytes: &[u8], endian: E) -> Option<Self> {
+        let bytes: &[u8; Self::WIDTH] = bytes[..Self::WIDTH].try_into().ok()?;
+        if endian.is_big_endian() {
+            Some(Self::from_be_bytes(*bytes))
+        } else {
+            Some(Self::from_le_bytes(*bytes))
+        }
     }
 }
 
-impl ReadBytes for u64 {
+impl RegisterValue for u64 {
     const WIDTH: usize = 8;
-    fn read_bytes<B: ByteOrder>(bytes: &[u8]) -> Option<Self> {
-        (bytes.len() >= Self::WIDTH).then(|| B::read_u64(bytes))
+    fn read_bytes<E: Endianness>(bytes: &[u8], endian: E) -> Option<Self> {
+        let bytes: &[u8; Self::WIDTH] = bytes[..Self::WIDTH].try_into().ok()?;
+        if endian.is_big_endian() {
+            Some(Self::from_be_bytes(*bytes))
+        } else {
+            Some(Self::from_le_bytes(*bytes))
+        }
     }
 }
 
-impl ReadBytes for i8 {
-    const WIDTH: usize = 1;
-    fn read_bytes<B: ByteOrder>(bytes: &[u8]) -> Option<Self> {
-        bytes.first().map(|b| *b as _)
-    }
+/// Provides access to a region of memory.
+pub trait MemoryRegion {
```

Comment:
Do we really need this trait? Looks like the only implementor will be `MemorySlice`, so we can just implement the function directly on that type. 

Having a trait could be nice if you expect other data to be passed in, but considering how simple `MemorySlice` is, you can just have a public constructor and users can always build that.

Reply:
The only other implementor is the `FakeMemoryRegion` struct in the `test` module. That can probably be replaced with `MemorySlice`s.

Reply:
Update: Rewriting the intermediate test to use a `MemorySlice` instead of the `FakeMemoryRegion` is actually *severely* annoying.